cmd/portforward.go

```diff
@@ -0,0 +1,548 @@
+				sendStopMessage = false
+				return
+			} else if m.Header.Proto == ProtoTypePortForward && m.Header.MsgType == MessageTypePortForward {
+				_, _ = p.conn.Write(m.Body)
+			}
+		case data := <-dataChan:
+			m := &ws.ProtoMsg{
+				Header: ws.ProtoHdr{
+					Proto:   ProtoTypePortForward,
+					MsgType: MessageTypePortForward,
+					Properties: map[string]interface{}{
+						propertyConnectionID: connectionID,
+					},
+				},
+				Body: data,
+			}
+			msgChan <- m
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// Run executes the command
+func (c *PortForwardCmd) Run() error {
+	ctx, cancelContext := context.WithCancel(context.Background())
+	defer cancelContext()
+
+	// get the JWT token
+	token, err := c.getToken()
+	if err != nil {
+		return err
+	}
+
+	// connect to the websocket and start the ping-pong connection health-check
+	client := deviceconnect.NewClient(c.server, c.skipVerify)
+	err = client.Connect(c.deviceID, token)
+	if err != nil {
+		return err
+	}
+
+	go client.PingPong(ctx)
+	defer client.Close()
+
+	// message channel
+	msgChan := make(chan *ws.ProtoMsg)
+
+	// start the local TCP listeners
+	for _, portMapping := range c.portMappings {
+		switch portMapping.Protocol {
+		case protocolTCP:
+			forwarder, err := NewTCPPortForwarder(c.bindingHost, portMapping.LocalPort,
+				portMapping.RemoteHost, portMapping.RemotePort)
+			if err != nil {
+				return err
+			}
+			go forwarder.Run(ctx, msgChan, c.connections)
+		case protocolUDP:
+			forwarder, err := NewUDPPortForwarder(c.bindingHost, portMapping.LocalPort,
+				portMapping.RemoteHost, portMapping.RemotePort)
+			if err != nil {
+				return err
+			}
+			go forwarder.Run(ctx, msgChan, c.connections)
+		default:
+			return errors.New("unknown protocol: " + portMapping.Protocol)
+		}
+	}
+
+	c.running = true
+	go c.processIncomingMessages(msgChan, client)
+
+	// handle CTRL+C and signals
+	quit := make(chan os.Signal, 1)
+	signal.Notify(quit, unix.SIGINT, unix.SIGTERM)
+
+	// wait for CTRL+C, signals or stop
+	healthcheckTimeout := time.Now().Add(24 * time.Hour)
+	for c.running {
+		select {
+		case msg := <-msgChan:
+			err := client.WriteMessage(msg)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error: %v\n", err)
+				break
+			}
+		case <-time.After(time.Until(healthcheckTimeout)):
+			c.err = errors.New("health check failed, connection with the device lost")
+			c.running = false
+		case <-quit:
+			c.running = false
+		case <-c.stop:
+			c.running = false
+		}
+	}
+
+	// cancel the context
+	cancelContext()
+
+	// return the error message (if any)
+	return c.err
+}
+
+func (c *PortForwardCmd) Stop() {
+	c.running = false
+	c.stop <- struct{}{}
+}
+
+func (c *PortForwardCmd) processIncomingMessages(msgChan chan *ws.ProtoMsg, client *deviceconnect.Client) {
+	for c.running {
+		m, err := client.ReadMessage()
+		if err != nil {
+			if c.running {
+				fmt.Fprintf(os.Stderr, "error: %v\n", err)
+			} else {
+				c.Stop()
+			}
```

Comment:
I do not get it, if it is running, then we output error, and in other case we stop?

Reply:
yes, we receive a spurious error when we stop (e.g. io.EOF), and we simply ignore it in this case